### PR TITLE
Use gRPC to handle communication and data transmission between object manager

### PR DESCRIPTION
**Pull request by jiangzihao2009**
## What do these changes do?
1. Remove object manager  send connection and accept socket, use gRPC to transfer data, send request, receive request(Pull, Push, FreeObjects).
2. Receive actual data in gRPC thread, substitute send_service and receive_service io for a rpc service.
3. Rpc service is a multi-threads service and all grpc send requests and receive requests are handled in this service. The object manager functions Pull/Push/FreeObjects are invoked in main service, however, the grpc service call(send request) is invoked in the rpc service. So no lock is needed except profile_events_ variable which is called in raylet main service.

## Related issue number
#4968 

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by raulchen** [comment]
cc @zhijunfu

File: src/ray/raylet/raylet.cc [hunk 1]
```diff
@@ -33,10 +33,6 @@ static const std::vector<std::string> node_manager_message_enum =
     GenerateEnumNames(ray::protocol::EnumNamesMessageType(),
```
Comment (zhijunfu) [inline src/ray/raylet/raylet.cc]: should the `node_manager_message_enum` be removed as well?

File: src/ray/rpc/grpc_server.cc [hunk 2]
```diff
@@ -27,8 +27,7 @@ void GrpcServer::Run() {
     }
   }
   // Start a thread that polls incoming requests.
-  std::thread polling_thread(&GrpcServer::PollEventsFromCompletionQueue, this);
-  polling_thread.detach();
```
Comment (zhijunfu) [inline src/ray/rpc/grpc_server.cc]: is detached() needed?

**Review by zhijunfu — COMMENTED** [review]
I left a few comments. Also could you do some testing to make sure there are no perf regressions with this change?

File: BUILD.bazel [hunk 3]
```diff
@@ -17,21 +37,42 @@ grpc_proto_library(
...
 )
 
+grpc_proto_library(
+    name = "object_manager_grpc_lib",
+    srcs = ["src/ray/protobuf/object_manager.proto"],
+)
+
+# Object manger rpc lib
+cc_library(
+    name = "object_manager_rpc_lib",
```
[the remark this answers is not in the document]
Comment (raulchen) [reply BUILD.bazel]: It seems that we cannot combine them. The `grpc_proto_library` rule needs to be defined separately.

File: BUILD.bazel [hunk 4]
```diff
@@ -8,6 +8,26 @@ load("@//bazel:cython_library.bzl", "pyx_library")
...
+cc_library(
+    name = "common_grpc_lib",
+    srcs = [
+        "src/ray/rpc/grpc_server.cc",
+    ],
+    hdrs = [
+        "src/ray/rpc/client_call.h",
+        "src/ray/rpc/grpc_server.h",
+        "src/ray/rpc/server_call.h",
+        "src/ray/rpc/util.h",
```
Comment (raulchen) [inline BUILD.bazel]: I think we keep the common rpc files under `rpc` folder, and move `node_manager/object_manager` rpc files to `rpc/node_manager` and `rpc/object_manager`. then you don't need to list all the files here. You can simply use `glob`.

File: python/ray/tests/test_basic.py [hunk 5]
```diff
@@ -2992,3 +2992,16 @@ def export_definitions_from_worker(remote_function, actor_class):
         ray.get(actor_handle.method.remote())
 
     ray.get(export_definitions_from_worker.remote(f, Actor))
+
+
+def test_forward(ray_start_cluster):
```
Comment (raulchen) [inline python/ray/tests/test_basic.py]: this test was used for debugging. not needed any more

File: src/ray/raylet/raylet.cc (hunk 1 above)
Comment (jiangzihao2009) [reply src/ray/raylet/raylet.cc]: raylet client hasn't changed, would remove it when change raylet client

File: src/ray/object_manager/object_manager.h [hunk 6]
```diff
@@ -67,7 +70,57 @@ class ObjectManagerInterface {
 };
 
 // TODO(hme): Add success/failure callbacks for push and pull.
-class ObjectManager : public ObjectManagerInterface {
+class ObjectManager : public ObjectManagerInterface,
+                      public rpc::ObjectManagerServiceHandler {
+  /// Implementation of object manager service
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]:
```suggestion
  /// Implementation of `ObjectManagerServiceHandler`
```
and put this under `public:`?

File: src/ray/rpc/grpc_server.cc [hunk 7]
```diff
@@ -48,7 +47,13 @@ void GrpcServer::PollEventsFromCompletionQueue() {
         // incoming request.
         server_call->GetFactory().CreateCall();
         server_call->SetState(ServerCallState::PROCESSING);
-        main_service_.post([server_call] { server_call->HandleRequest(); });
+        if (!main_service_.stopped()) {
+          main_service_.post([server_call] { server_call->HandleRequest(); });
+        } else {
+          // TODO: should send reply here
```
Comment (raulchen) [inline src/ray/rpc/grpc_server.cc]: Can you finish this TODO in this PR? The change should be simple. Just call `SendReply` with an error status. 
(maybe we should rename `SendReply` to `Finish`)

File: src/ray/object_manager/object_manager.cc [hunk 8]
```diff
@@ -30,35 +31,25 @@ ObjectManager::ObjectManager(asio::io_service &main_service,
       });
   store_notification_.SubscribeObjDeleted(
       [this](const ObjectID &oid) { NotifyDirectoryObjectDeleted(oid); });
-  StartIOService();
+
+  StartRpcService();
+  object_manager_server_.Run();
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.cc]: move `object_manager_server_.Run();` to `StartRpcService`?

File: src/ray/object_manager/object_manager.cc [hunk 9]
```diff
@@ -238,51 +232,18 @@ void ObjectManager::TryPull(const ObjectID &object_id) {
...
-  conn->WriteMessageAsync(
-      static_cast<int64_t>(object_manager_protocol::MessageType::PullRequest),
-      fbb.GetSize(), fbb.GetBufferPointer(), [this, conn](ray::Status status) {
-        if (!status.ok()) {
-          RAY_CHECK(status.IsIOError())
-              << "Failed to contact remote object manager during Pull";
-          connection_pool_.RemoveSender(conn);
-        }
-      });
+  auto callback = [](const Status &status, const rpc::PullReply &reply) {};
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.cc]: just pass in `nullptr`

File: src/ray/object_manager/object_manager.cc [hunk 10]
```diff
@@ -708,194 +627,84 @@ void ObjectManager::WaitComplete(const UniqueID &wait_id) {
...
-  } else {
-    // Close the connection by skipping the call to ProcessMessages.
-    RAY_LOG(ERROR) << "Failed to ExecuteReceiveObject from remote object manager, error: "
-                   << status;
+void ObjectManager::HandlePullRequest(const rpc::PullRequest &request,
+                                      rpc::PullReply *reply,
+                                      rpc::RequestDoneCallback done_callback) {
+  ObjectID object_id = ObjectID::FromBinary(request.object_id());
+  ClientID client_id = ClientID::FromBinary(request.client_id());
+  RAY_LOG(DEBUG) << "Receive pull request from client " << client_id << " for object ["
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.cc]:
```suggestion
  RAY_LOG(DEBUG) << "Received pull request from client " << client_id << " for object ["
```

File: src/ray/object_manager/object_manager.cc [hunk 11]
```diff
@@ -708,194 +627,84 @@ void ObjectManager::WaitComplete(const UniqueID &wait_id) {
...
-  FreeObjects(object_ids, local_only);
-  conn->ProcessMessages();
+void ObjectManager::HandleFreeObjectsRequest(const rpc::FreeObjectsRequest &request,
+                                             rpc::FreeObjectsReply *reply,
+                                             rpc::RequestDoneCallback done_callback) {
+  std::vector<ObjectID> object_ids;
+  for (const auto &e : request.object_ids()) {
+    object_ids.emplace_back(ObjectID::FromBinary(e));
+  }
+  FreeObjects(object_ids, true /* local_only */);
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.cc]:
```suggestion
  FreeObjects(object_ids, /* local_only = */ true);
```

File: src/ray/object_manager/object_manager.cc [hunk 12]
```diff
@@ -908,47 +717,55 @@ void ObjectManager::FreeObjects(const std::vector<ObjectID> &object_ids,
...
+    RemoteConnectionInfo connection_info(client_id);
+    object_directory_->LookupRemoteConnectionInfo(connection_info);
+    if (!connection_info.Connected()) {
+      return nullptr;
+    }
+    auto object_manager_client = std::make_shared<rpc::ObjectManagerClient>(
+        connection_info.ip, connection_info.port, client_call_manager_);
+
+    if (!object_manager_client) {
+      return nullptr;
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.cc]: why would `object_manager_client` be nullptr here?

File: src/ray/object_manager/object_manager.h (hunk 6 above)
Comment (jiangzihao2009) [reply src/ray/object_manager/object_manager.h]: It's not only handler, and some other func for rpc service

File: src/ray/rpc/grpc_server.cc (hunk 7 above)
Comment (jiangzihao2009) [reply src/ray/rpc/grpc_server.cc]: ok, i will chang it to Finish and invoke here

File: src/ray/object_manager/object_manager.cc [hunk 13]
```diff
@@ -481,36 +417,17 @@ ray::Status ObjectManager::SendObjectHeaders(const UniqueID &push_id,
...
-      fbb.GetSize(), fbb.GetBufferPointer());
-  if (!status.ok()) {
-    return status;
-  }
-  return SendObjectData(object_id, chunk_info, conn);
-}
+  std::string buffer;
+  buffer.resize(chunk_info.buffer_length);
+  buffer.assign(chunk_info.data, chunk_info.data + chunk_info.buffer_length);
+  push_request.set_data(std::move(buffer));
```
[the remark this answers is not in the document]
Comment (jiangzihao2009) [reply src/ray/object_manager/object_manager.cc]: i plan to change rpc_service to multithread and add a multi-thread send_service for SendObjectChunk

File: src/ray/object_manager/object_manager.h [hunk 14]
```diff
@@ -361,28 +405,19 @@ class ObjectManager : public ObjectManagerInterface {
   ObjectStoreNotificationManager store_notification_;
   ObjectBufferPool buffer_pool_;
 
-  /// This runs on a thread pool dedicated to sending objects.
-  boost::asio::io_service send_service_;
-  /// This runs on a thread pool dedicated to receiving objects.
-  boost::asio::io_service receive_service_;
+  /// Multi-thread asio service, deal with all send and receive request
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]:
```suggestion
  /// Multi-thread asio service, used to deal with all outgoing and incoming RPC requests from/to remote object managers.
```

File: src/ray/object_manager/object_manager.h [hunk 15]
```diff
@@ -361,28 +405,19 @@ class ObjectManager : public ObjectManagerInterface {
...
+  boost::asio::io_service rpc_service_;
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]: put this together with `rpc_work` and `rpc_threads_`?

File: src/ray/object_manager/object_manager.h [hunk 16]
```diff
@@ -361,28 +405,19 @@ class ObjectManager : public ObjectManagerInterface {
...
-  boost::asio::io_service::work receive_work_;
+  boost::asio::io_service::work rpc_work_;
 
-  /// Runs the send service, which handle
-  /// all outgoing object transfers.
-  std::vector<std::thread> send_threads_;
-  /// Runs the receive service, which handle
-  /// all incoming object transfers.
-  std::vector<std::thread> receive_threads_;
+  /// Runs the rpc service, which handle all send request and receive request
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]:
```suggestion
  /// The thread pool used for running `rpc_service_`. 
```
`rpc_service_`'s documentation already explain what it does.

File: src/ray/object_manager/object_manager.h [hunk 17]
```diff
@@ -67,7 +73,81 @@ class ObjectManagerInterface {
 };
 
 // TODO(hme): Add success/failure callbacks for push and pull.
-class ObjectManager : public ObjectManagerInterface {
+class ObjectManager : public ObjectManagerInterface,
+                      public rpc::ObjectManagerServiceHandler {
+ public:
+  /// Implementation of object manager service
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]: Nit. every comment should end with a period.

File: src/ray/object_manager/test/object_manager_stress_test.cc [hunk 18]
```diff
@@ -129,7 +99,7 @@ class TestObjectManagerBase : public ::testing::Test {
     int max_receives_a = 2;
     int max_sends_b = 3;
     int max_receives_b = 3;
-    uint64_t object_chunk_size = static_cast<uint64_t>(std::pow(10, 3));
+    uint64_t object_chunk_size = static_cast<uint64_t>(std::pow(10, 5));
```
Comment (raulchen) [inline src/ray/object_manager/test/object_manager_stress_test.cc]: Is this change needed for this PR?

File: src/ray/protobuf/object_manager.proto [hunk 19]
```diff
@@ -0,0 +1,65 @@
+syntax = "proto3";
+
+package ray.rpc;
+
+message ObjectInfo {
```
Comment (raulchen) [inline src/ray/protobuf/object_manager.proto]: Do we need to define `ObjectInfo` in pb? 
I think it's only used between plasma and raylet, right?

File: src/ray/protobuf/object_manager.proto [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+}
+message PullReply {
+}
+message FreeObjectsReply {
+}
+
+service ObjectManagerService {
+  rpc Push(PushRequest) returns (PushReply);
+  rpc Pull(PullRequest) returns (PullReply);
+  rpc FreeObjects(FreeObjectsRequest) returns (FreeObjectsReply);
```
Comment (raulchen) [inline src/ray/protobuf/object_manager.proto]: document this service and its interfaces

File: src/ray/rpc/grpc_server.cc [hunk 21]
```diff
@@ -48,7 +47,15 @@ void GrpcServer::PollEventsFromCompletionQueue() {
...
         server_call->SetState(ServerCallState::PROCESSING);
-        main_service_.post([server_call] { server_call->HandleRequest(); });
+        if (!main_service_.stopped()) {
+          main_service_.post([server_call] { server_call->HandleRequest(); });
+        } else {
+          // Handle service for rpc call has stopped, we must handle the call here
+          // to send reply and remove it from cq
+          RAY_LOG(DEBUG) << "Handle service has been closed.";
+          server_call->Finish(Status::Invalid("HandleServiceClosed"));
+          delete_call = true;
```
Comment (raulchen) [inline src/ray/rpc/grpc_server.cc]: should we delete this call here? 
I think when the reply is sent, another event will be triggered. we should delete this call at that time.

File: src/ray/object_manager/object_manager.h [hunk 22]
```diff
@@ -49,6 +52,9 @@ struct ObjectManagerConfig {
   /// Negative: waiting infinitely.
   /// 0: giving up retrying immediately.
   int push_timeout_ms;
+  /// Number of threads of rpc service
+  /// Send and receive request in these threads
+  int rpc_service_threads_number;
```
Comment (raulchen) [inline src/ray/object_manager/object_manager.h]: `max_sends` and `max_receives` are not needed any more?

**Review by raulchen — COMMENTED** [review]
Looks good overall! I just left some small comments.

File: BUILD.bazel (hunk 4 above)
Comment (zhijunfu) [reply BUILD.bazel]: +1 for moving these files

File: src/ray/object_manager/object_manager.cc [hunk 23]
```diff
@@ -708,247 +638,150 @@ void ObjectManager::WaitComplete(const UniqueID &wait_id) {
...
+    std::vector<std::shared_ptr<rpc::ObjectManagerClient>> rpc_clients;
+    for (const auto &connection_info : remote_connections) {
+      auto rpc_client = GetRpcClient(connection_info.client_id);
+      if (rpc_client != nullptr) {
+        rpc_clients.push_back(rpc_client);
+      }
+    }
+    rpc_service_.post([this, object_ids, rpc_clients]() {
+      SpreadFreeObjectsRequest(object_ids, rpc_clients);
+    });
```
[the remark this answers is not in the document]
Comment (jiangzihao2009) [reply src/ray/object_manager/object_manager.cc]: SpreadFreeObjectsRequest is run in multi thread, the object manager client hash map can only handled in main thread to avoid lock

File: src/ray/rpc/grpc_server.cc (hunk 21 above)
Comment (jiangzihao2009) [reply src/ray/rpc/grpc_server.cc]: yes, double free may occur here!

File: src/ray/object_manager/object_manager.h [hunk 24]
```diff
@@ -67,7 +73,81 @@ class ObjectManagerInterface {
...
+  /// Handle push request from remote object manager
+  ///
+  /// Push request will contain the object which is specified by pull request
+  /// the object will be transfered by a sequence of chunks.
+  ///
+  /// \param request Push request including the object chunk data
+  /// \param reply Reply to the sender
+  /// \param done_callback Callback of the request
+  void HandlePushRequest(const rpc::PushRequest &request, rpc::PushReply *reply,
+                         rpc::RequestDoneCallback done_callback);
```
Comment (zhijunfu) [inline src/ray/object_manager/object_manager.h]:
```suggestion
                         rpc::RequestDoneCallback done_callback) override;
```

File: src/ray/object_manager/object_manager.h [hunk 25]
```diff
@@ -67,7 +73,81 @@ class ObjectManagerInterface {
...
+
+  /// Handle pull request from remote object manager
+  ///
+  /// \param request Pull request
+  /// \param reply Reply
+  /// \param done_callback Callback of request
+  void HandlePullRequest(const rpc::PullRequest &request, rpc::PullReply *reply,
+                         rpc::RequestDoneCallback done_callback);
```
Comment (zhijunfu) [inline src/ray/object_manager/object_manager.h]:
```suggestion
                         rpc::RequestDoneCallback done_callback) override;
```

File: src/ray/object_manager/object_manager.h [hunk 26]
```diff
@@ -67,7 +73,81 @@ class ObjectManagerInterface {
...
+
+  /// Handle free objects request
+  ///
+  /// \param request Free objects request
+  /// \param reply Reply
+  /// \param done_callback
+  void HandleFreeObjectsRequest(const rpc::FreeObjectsRequest &request,
+                                rpc::FreeObjectsReply *reply,
+                                rpc::RequestDoneCallback done_callback);
```
Comment (zhijunfu) [inline src/ray/object_manager/object_manager.h]:
```suggestion
                                rpc::RequestDoneCallback done_callback) override;
```

**Review by zhijunfu — COMMENTED** [review]
Overall looks good. 

I would suggest to add a comment section in object_manager.h on how the thread is implemented. 
And also it would be good to mark all function declarations which are executed in `rpc service` with a note so that people are aware which context this function is supposed to run.

Also I think we should add some performance test to make sure there's no performance regression. For example, a simple test can be to continuously submit tasks to an actor with by reference arguments.

File: BUILD.bazel [hunk 27]
```diff
@@ -287,6 +328,7 @@ cc_library(
     deps = [
         ":gcs",
         ":object_manager_fbs",
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply BUILD.bazel]: It would be ok to move it into https://github.com/ray-project/ray/blob/master/src/ray/raylet/format/node_manager.fbs if you want to. I guess eventually everything will be converted to protobuf so we can also do it later.

File: src/ray/object_manager/object_manager.cc [hunk 28]
```diff
@@ -201,8 +197,18 @@ void ObjectManager::TryPull(const ObjectID &object_id) {
...
   RAY_LOG(DEBUG) << "Sending pull request from " << client_id_ << " to " << client_id
                  << " of object " << object_id;
-  // Try pulling from the client.
-  PullEstablishConnection(object_id, client_id);
+
+  auto rpc_client = GetRpcClient(client_id);
+  if (rpc_client) {
+    // Try pulling from the client.
+    rpc_service_.post([this, object_id, client_id, rpc_client]() {
+      auto st = SendPullRequest(object_id, client_id, rpc_client);
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.cc]: this status should be checked

File: src/ray/rpc/object_manager/object_manager_server.h [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace ray {
+namespace rpc {
+
+/// Implementations of the `ObjectManagerService`, check interface in
+/// `src/ray/protobuf/object_manager.proto`.
+class ObjectManagerServiceHandler {
+ public:
+  /// Handle a `Push` request.
+  /// The implementation can handle this request asynchronously. When hanling is done, the
```
Comment (pcmoritz) [inline src/ray/rpc/object_manager/object_manager_server.h]: hanling -> handling

File: src/ray/rpc/object_manager/object_manager_server.h [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+                                        RequestDoneCallback done_callback) = 0;
+};
+
+/// The `GrpcServer` for `ObjectManagerService`.
+class ObjectManagerServer : public GrpcServer {
+ public:
+  /// Construct a `ObjectManagerServer`.
+  ///
+  /// \param[in] port See `GrpcServer`.
+  /// \param[in] handler The service handler that actually handle the requests.
```
Comment (pcmoritz) [inline src/ray/rpc/object_manager/object_manager_server.h]: you mean "service_handler" here, right?

File: src/ray/rpc/client_call.h [hunk 31]
```diff
@@ -47,7 +47,11 @@ using ClientCallback = std::function<void(const Status &status, const Reply &rep
 template <class Reply>
 class ClientCallImpl : public ClientCall {
  public:
-  void OnReplyReceived() override { callback_(GrpcStatusToRayStatus(status_), reply_); }
+  void OnReplyReceived() override {
+    if (callback_ != nullptr) {
```
Comment (pcmoritz) [inline src/ray/rpc/client_call.h]: when would the `callback_` be null?

Comment (jiangzihao2009) [reply src/ray/rpc/client_call.h]: It's user input param, user can enter null

File: src/ray/object_manager/object_manager.cc (hunk 28 above)
Comment (jiangzihao2009) [reply src/ray/object_manager/object_manager.cc]: i think it's always ok, the exception would be handled by client callback(at present, only a warning is shown)

File: src/ray/rpc/object_manager/object_manager_server.h (hunk 30 above)
Comment (jiangzihao2009) [reply src/ray/rpc/object_manager/object_manager_server.h]: yes

File: src/ray/object_manager/object_manager.cc (hunk 28 above)
Comment (raulchen) [reply src/ray/object_manager/object_manager.cc]: then make the function return void?

**Comment by raulchen** [comment]
Jenkins, retest this.

File: BUILD.bazel [hunk 32]
```diff
@@ -9,6 +9,24 @@ load("@//bazel:cython_library.bzl", "pyx_library")
 
 COPTS = ["-DRAY_USE_GLOG"]
 
+# grpc common lib
+cc_library(
```
Comment (raulchen) [inline BUILD.bazel]: put this after the protobuf definitions?

**Review by raulchen — APPROVED** [review]

**Comment by pcmoritz** [comment]
@raulchen This PR is breaking valgrind, can you have a look?

**Comment by raulchen** [comment]
> @raulchen This PR is breaking valgrind, can you have a look?

sorry about that, I'll take a look later today.
